federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+
+	toDistribute := replicaNums{
+		min: remainingReplicas.min + int32(clusterLists.availableMin.Len()),
+		max: remainingReplicas.max + int32(clusterLists.availableMax.Len()),
+	}
+
+	// Pass 2: Distribute Max and then Min.
+	// Here we first distribute max and then (in the next loop)
+	// distribute min into those clusters which already get the
+	// max fixed.
+	// In this process we might not meet the min limit and total of
+	// min limits might remain more then the requested federated min.
+	// This is partially because a min per cluster cannot be lesser
+	// then 1, but min could be requested as 1 at federation.
+	// Additionally we first increase replicas into those clusters
+	// which already have hpa's and are in a condition to increase.
+	// This will save cluster related resources for the user, such that
+	// if an already existing cluster can satisfy users request why send
+	// the workload to another.
+	// We then go ahead to give the replicas to those which do not
+	// have any hpa. In this pass however we try to ensure that all
+	// our Max are consumed in this reconcile.
+	toDistribute.max = distributeMaxReplicas(toDistribute.max, clusterLists, rdc, currentObjs, scheduleStatus)
+
+	// We distribute min to those clusters which:
+	// 1 - can adjust min (our increase step would be only 1)
+	// 2 - which do not have this hpa and got max(increase step rdcMin)
+	// We might exhaust all min replicas here, with
+	// some clusters still needing them. We adjust this in finalise by
+	// assigning min replicas to 1 into those clusters which got max
+	// but min remains 0.
+	toDistribute.min = distributeMinReplicas(toDistribute.min, clusterLists, rdc, currentObjs, scheduleStatus)
+
+	err := a.FedUpdateStatus(fedObj, currentObjs)
+	if err != nil {
+		return nil, err
+	}
+	// Pass 3: finalise
+	return a.finaliseScheduled(fedObj, scheduleStatus)
+}
+
+func (a *HpaAdapter) FedUpdateStatus(obj pkgruntime.Object, currentObjs map[string]pkgruntime.Object) error {
+	copiedObj, err := api.Scheme.DeepCopy(obj)
+	if err != nil {
+		return fmt.Errorf("Error copying federated hpa object: %v", err)
+	}
+	if len(currentObjs) == 0 {
+		return nil
+	}
+
+	fedHpa := copiedObj.(*autoscalingv1.HorizontalPodAutoscaler)
+	gen := fedHpa.Generation
+	updatedStatus := autoscalingv1.HorizontalPodAutoscalerStatus{ObservedGeneration: &gen}
+	if fedHpa.Status.LastScaleTime != nil {
+		t := metav1.NewTime(fedHpa.Status.LastScaleTime.Time)
+		updatedStatus.LastScaleTime = &t
+	}
+	var ccup int32 = 0
+	updatedStatus.CurrentCPUUtilizationPercentage = &ccup
+
+	count := int32(0)
+	for _, lObj := range currentObjs {
+		if lObj == nil {
+			continue
+		}
+		lHpa := lObj.(*autoscalingv1.HorizontalPodAutoscaler)
+		if lHpa.Status.CurrentCPUUtilizationPercentage != nil {
+			*updatedStatus.CurrentCPUUtilizationPercentage += *lHpa.Status.CurrentCPUUtilizationPercentage
+			count++
+		}
+		if updatedStatus.LastScaleTime != nil && lHpa.Status.LastScaleTime != nil &&
```

Comment:
Yes, updated to a better condition.
```
		if clusterHpa.Status.LastScaleTime != nil {
			t := metav1.NewTime(clusterHpa.Status.LastScaleTime.Time)
			if typedInfo.fedStatus.lastScaleTime != nil &&
				t.After(typedInfo.fedStatus.lastScaleTime.Time) {
				typedInfo.fedStatus.lastScaleTime = &t
			}
		}
```

clusterHpa is the cluster local object being observed
typedInfo.fedStatus holds the latest scaled time compared and updated till now (being compared with local objects one by one, and the latest scale time existing in fed status)